state/modeltransaction.go

```diff
@@ -0,0 +1,83 @@
+// Copyright 2017 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package state
+
+import (
+	"github.com/juju/errors"
+	jujutxn "github.com/juju/txn"
+	"gopkg.in/mgo.v2/txn"
+)
+
+// ModelTransaction is a high-level transaction for making changes to a model.
+//
+// NOTE(axw) when State methods and entities are moved to Model, then this
+// should encapsulate a Model rather than a State, and NewModelTransaction
+// should be moved to Model and renamed NewTransaction.
+type ModelTransaction struct {
+	st  *State
+	ops []ModelTransactionOperation
+}
+
+// ModelTransactionOperation is a high-level transaction operation,
+// encapsulating the logic required to apply a change to a model.
+type ModelTransactionOperation interface {
+	// Build builds the low-level database transaction operations required
+	// to apply the change. If the transaction operations fail (e.g. due
+	// to concurrent changes), then Build may be called again. The attempt
+	// number can be obtained from the ModelTransactionContext.
```

Comment:
Perhaps document the special errors that Build can return (for example, I see a test for ErrNoOperations below) ?